### ISSUE TITLE
jet for mixed partial derivatives

### ISSUE THREAD

**Issue by vduarte** [google/jax#5152]
@mattjj @jessebett 
Hi all,

I was wondering if there is a way of using the jet API to extract partial derivatives. I know they could be computed by interpolating the univariate Taylor series, but I wonder if there is a more straightforward way. The nested version of my problem is something like this:
```
from jax import jacfwd
import jax.numpy as jnp

def f(x1, x2, x3, x4):
  return jnp.sin(x1**2 * x2**2 * x3**2 * x4**2)


primals = (.1, .2, .3, .4)
jacfwd(jacfwd(jacfwd(jacfwd(f, 0), 1), 2), 3)(*primals)
```

In Julia, the package TaylorSeries solves this problem very easily:

```
using TaylorSeries, Plots

function f(x1, x2, x3, x4)
  return sin((.1 + dx1)^2 * (.2 + dx2)^2 * (.3 + dx3)^2 * (.4 + dx4)^2)
end

dx1, dx2, dx3, dx4  = set_variables("dx1, dx2, dx3, dx4", order=4)
polynomial = f(x1, x2, x3, x4)
print(getcoeff(polynomial, [1, 1, 1, 1]))
```

Is there a way to do something similar with jet? Many thanks!

**Comment by mattjj** [comment]
Thanks for the question!

Hrm, I don't _think_ this is any more efficient using `jet` compared with just using `jvp` (essentially `jacfwd`) in your example. That is, what distinguishes `jet` from just composing up `jvp` is that with `jvp` we can compute things of the form ∂^k f(x)[v1, v2, ..., vk], whereas `jet` is specialized to the case of computing things of the form ∂^k f(x)[v, v, ..., v], leveraging the fact that we contract against many copies of the same vector (as one does in a Taylor series).

Here's a snippet from a pdf we wrote about this:

![image](https://user-images.githubusercontent.com/1458824/101698324-234baa00-3a2e-11eb-9733-8eb3f4137023.png)

AIUI in your example you want to evaluate ∂^4 f(x)[e1, e2, e3, e4], where ek is the kth standard basis vector. Since we're applying ∂^4 f(x) to four different vectors, I don't think we fall into the case where Taylor series propagation is helpful.

WDYT? Does this sound right? I'm kinda rusty at this stuff and would appreciate checks / pushback / discussion.

**Comment by vduarte** [comment]
Hi Matt,

Thanks for the quick reply and the reference! My understanding is that jet could be helpful if we choose the perturbation directions carefully and then combine these results to produce the desired partial derivatives. 
Here is a concrete example. Let me rewrite the problem as a 3D problem instead of 4D:

```
from jax import jacfwd, jit
from jax.experimental.jet import jet
import jax.numpy as jnp


def f(x1, x2, x3):
  return jnp.sin(x1**2 * x2**2 * x3**2)

primals = (.1, .2, .3)
print(jacfwd(jacfwd(jacfwd(f, 0), 1), 2)(*primals))
```

We could compute this partial derivative with a single jet call:

```
series = [[jnp.array([1., 1., 1., 1.]), 0., 0.],
          [jnp.array([1., -1., -1., 1]), 0., 0.],
          [jnp.array([1., -1., 1., -1]), 0., 0.]]

d3 = jet(f, primals, series)[1][2]
f111 = (d3[0] + d3[1] - d3[2] - d3[3]) / 24
print(f111)
```
As you mentioned, in this case, the nested version is faster than the jet one. The problem is that in my application I actually have more than 15 variables, and memory becomes an issue. My hope is that jet would do better in that case. 

I had the impression that jet was able to compute mixed partial derivatives out of the box (I got this impression from Griewank's comment here: [(https://openreview.net/forum?id=SkxEF3FNPH)].

Thanks again for the help!

**Comment by thisiscam** [comment]
Hi, I wonder if @vduarte have  found an answer/solution to this?

**Comment by DavidLanders95** [comment]
I too would like to know if there is an answer to this issue.

**Comment by mattjj** [comment]
> My understanding is that jet could be helpful if we choose the perturbation directions carefully and then combine these results to produce the desired partial derivatives

Yes, I suspect that's true: one of the authors of https://openreview.net/pdf?id=J2wI2rCG2u mentioned this recently as well. That paper might include some information about it, but I don't have time to read it at the moment, so I thought I'd just leave it here as a potential clue. Sorry I don't have the answer at the moment.

**Comment by mattjj** [comment]
Maybe in Appendix F of that paper?

**Comment by thisiscam** [comment]
I was actually able to find the right technique in the classic AD book. See relevant discussion here for the references https://github.com/sail-sg/stde/issues/1.

**Comment by thisiscam** [comment]
For anyone interested, I've shared my old implementation here: https://gist.github.com/thisiscam/7b77c18d8dcdc34d0404cc575a74a8ce. I tested it on some random inputs and functions, and it seems to be working correctly.

Currently, this implementation handles the case of computing a single mixed partial derivative (following Eq. 13 from [GUW98](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=eb1825fcd4fcd64df4d3693390cd2ec0ad8b2b33)). If your use case requires computing multiple mixed partials, there's room for optimization by leveraging shared computation, as described in Eq. 17 of GUW98.

Hope this is helpful!

**Comment by DavidLanders95** [comment]
> For anyone interested, I've shared my old implementation here: https://gist.github.com/thisiscam/7b77c18d8dcdc34d0404cc575a74a8ce. I tested it on some random inputs and functions, and it seems to be working correctly.
> 
> Currently, this implementation handles the case of computing a single mixed partial derivative (following Eq. 13 from [GUW98](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=eb1825fcd4fcd64df4d3693390cd2ec0ad8b2b33)). If your use case requires computing multiple mixed partials, there's room for optimization by leveraging shared computation, as described in Eq. 17 of GUW98.
> 
> Hope this is helpful!

Aw, amazing! Thank you very much, it's beautiful. I feel like this example couldbe cited in the documentation, it's a great demonstration of the power of jet.

**Comment by thisiscam** [comment]
> > For anyone interested, I've shared my old implementation here: https://gist.github.com/thisiscam/7b77c18d8dcdc34d0404cc575a74a8ce. I tested it on some random inputs and functions, and it seems to be working correctly.
> > Currently, this implementation handles the case of computing a single mixed partial derivative (following Eq. 13 from [GUW98](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=eb1825fcd4fcd64df4d3693390cd2ec0ad8b2b33)). If your use case requires computing multiple mixed partials, there's room for optimization by leveraging shared computation, as described in Eq. 17 of GUW98.
> > Hope this is helpful!
> 
> Ah, amazing! Thank you very much, it's beautiful. I feel like this example could be cited in the documentation, it's a great demonstration of the power of jet. It's simple enough to understand, and it's about 20 times faster on my machine than using jacfwd.

Thanks! Note that the naive jacfwd implementation in my code is not really the best though, so it's not a fair comparison! Ideally the naive impl should be doing `jacfwd`s one variable at a time (instead of the entire x0 at a time). If you try that please report back on the performance!
